include/exotic/cester.h

```diff
@@ -951,35 +951,46 @@ static __CESTER_INLINE__ void cester_str_value_after_first(char *arg, char from,
 }
 
 static __CESTER_INLINE__ void cester_concat_str(char **out, const char * extra) {
-    size_t concatted_pos = 0, index = 0;
-    char *concatted = (char*) malloc(sizeof(char) * 80000 );
-    if (extra == CESTER_NULL) {
-        extra = (char *) "(null)";
+
+    // if nothing to concatenate - do nothing
+    if (!extra) {
+        return;
     }
-    while (1) {
-        if ((*out) == CESTER_NULL || (*out)[index] == '\0') {
-            break;
-        }
-        concatted[concatted_pos] = (*out)[index];
-        concatted_pos++;
-        index++;
+
+    // no point doing anything if being asked to concatenate empty extra string
+    size_t extra_length = strlen(extra);
+    if (extra_length == 0) {
+        return;
     }
-    index = 0;
-    while (1) {
-        if (extra[index] == '\0') {
-            break;
-        }
-        concatted[concatted_pos] = extra[index];
-        concatted_pos++;
-        index++;
+
+    // get the length of the string being appended to
+    size_t original_length = 0;
+    if (*out) {
+        original_length = strlen(*out);
     }
-    concatted[concatted_pos] = '\0';
-    if (strlen(*out) > 0) {
+
+    // allocate the correct length
+    size_t new_length = original_length + extra_length;
+    char * concatted = (char *)malloc(sizeof(char) * new_length + 1);
+
+    // put the original string in the target
+    if (original_length > 0) {
+        strcpy(concatted, *out);
+    }
+    else {
+        concatted[0] = 0;
+    }
+
+    // append the extra string
+    strcat(concatted, extra);
+
+    // free the originals - if the string is empty, we assume
+    // that it is a non-malloced initializer and don't free it.
+    if (*out && *out[0] != 0) {
         free(*out);
     }
-    *out = (char*) malloc(concatted_pos+1);
-    cester_copy_str(&concatted, out, concatted_pos);
-    free(concatted);
+    // set the value
+    *out = concatted;
```

Comment:
The frame work is written to support early  C standards such as ANSI, C89. In those standard the C++ style single line comment `//` is not supported, hence you should move those inline comment out of the function and write it as a proper doc for the function before the function using the comment style `/** ... */`

Reply:
Also, ISO C90 forbids mixed declarations and code so you cannot have a new variable declare after writing some expression and code. All variables are to be declared at the beginning of the scope before any code, even if they are not initialized on the spot.

Reply:
The regression fails for string testing as I make an assumption that cester_concate_str will ALWAYS set the out parameter to a mallocated value. WIth your update that not the case any more hence any test case using cester_assert_str_equal will fail is one of it, parameter is not a mallocated value.

So this update will cause any test using cester_assert_str_equal to fail, the solution to that is to init `expected` and `actual` to CESTER_NULL in the function cester_evaluate_expect_actual_str.


~expected = (char*) "";~
~actual = (char*) "";~

```c
    expected = CESTER_NULL;
    actual = CESTER_NULL;
```

And check if it CESTER_NULL before attempting to free the two variables at the end of the function cester_evaluate_expect_actual_str.

~free(expected)~
~free(actual)~

```c
    if (expected != CESTER_NULL) {
        free(expected);
    }
    if (actual != CESTER_NULL) {
        free(actual);
    }
```

Reply:
The same version of your update compatible with ANSI and C89

```c
static __CESTER_INLINE__ void cester_concat_str(char **out, const char * extra) {
    char *concatted;
    size_t new_length;
    size_t extra_length;
    size_t original_length;
    if (!extra) {
        return;
    }

    extra_length = strlen(extra);
    if (extra_length == 0) {
        return;
    }

    original_length = 0;
    if (*out) {
        original_length = strlen(*out);
    }

    new_length = original_length + extra_length;
    concatted = (char *)malloc(sizeof(char) * new_length + 1);

    if (original_length > 0) {
        strcpy(concatted, *out);
    }
    else {
        concatted[0] = 0;
    }

    strcat(concatted, extra);

    if (*out && *out[0] != 0) {
        free(*out);
    }
    *out = concatted;
}
```

Reply:
My main reason for staying away from str* related function is for safety and less dependent on string.h, for safety purpose replace `strcpy` and `strcat` with it safe alternatives `strncpy` and `strncat`